contracts/implementations/nodes/SyntheticDevice.sol

```diff
@@ -82,6 +82,96 @@ contract SyntheticDevice is AccessControlInternal {
 
     // ***** Interaction with nodes *****//
 
+    /**
+     * @notice Mints a list of synthetic devices and link them to vehicles
+     * To be called for existing vehicles already connected
+     * @dev Caller must have the admin role
+     * @param data input data with the following fields:
+     *  integrationNode -> Parent integration node id
+     *  vehicleNode -> Vehicle node id
+     *  syntheticDeviceSig -> Synthetic Device's signature hash
+     *  syntheticDeviceAddr -> Address associated with the synthetic device
+     *  attrInfoPairs -> List of attribute-info pairs to be added
+     */
+    function mintSyntheticDeviceBatch(
+        Types.MintSyntheticDeviceBatchInput[] calldata data
+    ) external onlyRole(Roles.DEFAULT_ADMIN_ROLE) {
+        NodesStorage.Storage storage ns = NodesStorage.getStorage();
+        MapperStorage.Storage storage ms = MapperStorage.getStorage();
+        SyntheticDeviceStorage.Storage storage sds = SyntheticDeviceStorage
+            .getStorage();
+
+        address vehicleIdProxyAddress = VehicleStorage
+            .getStorage()
+            .idProxyAddress;
+        address sdIdProxyAddress = sds.idProxyAddress;
+
+        address owner;
+        bytes32 message;
+        uint256 newTokenId;
+        for (uint256 i = 0; i < data.length; i++) {
```

Comment:
There are a lot of repeated property and map accesses in here. Does it help readability (or gas, for that matter) to pull some of those up?

Reply:
Guess what. Stack too deep. If I add one variable more we hit the stack.

Reply:
Thanks for checking!